util/ranger/points.go

```diff
@@ -250,6 +250,36 @@ func (r *builder) buildFormBinOp(expr *expression.ScalarFunction) []point {
 		return nil
 	}
 
+	// if the colunm is unsigned int or unsigned decimal, and the value
+	// is less than zero, we only need compare with zero.
+	if mysql.HasUnsignedFlag(ft.Flag) {
+		needFix := false
+		zeroValue := new(types.Datum)
+		if mysql.IsIntegerType(ft.Tp) && value.Kind() == types.KindInt64 {
+			needFix = true
+			zeroValue.SetInt64(int64(0))
+		} else if (ft.Tp == mysql.TypeDecimal || ft.Tp == mysql.TypeNewDecimal) &&
+			value.Kind() == types.KindMysqlDecimal {
+			needFix = true
+			zeroValue.SetMysqlDecimal(new(types.MyDecimal))
+		}
```

Comment:
should we also consider float or double values?

Reply:
we don't need consider float and decimal, only unsigned integer will convert a negative number to positive number in function `func (d *Datum) convertToUint(sc *stmtctx.StatementContext, target *FieldType) (Datum, error)`